Versioned dom

asajeffrey opened the pull request:
This PR adds versioning to the DOM. There are now node.get_version and node.get_descendent_version methods that return a counter that is bumped when the node is dirtied. This is used to implement cache invalidation for caching HTMLCollection state. Caching HTMCollections gets a 1000x speedup in the Dromaeo DOM query tests.

Addresses https://github.com/servo/servo/issues/6901, https://github.com/servo/servo/issues/3381 and https://github.com/servo/servo/issues/1916.

Replaces PR https://github.com/servo/servo/pull/6927.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8227)

<!-- Reviewable:end -->

asajeffrey commented:
The PR in its current state needs squashing, which I'll do if it passes review.

eefriedman commented:
More comments would be nice.
<!-- Reviewable comment -K1fArDWu0iK_aNQq8G_:1813486277 -->

---

Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/document.rs, line 1214 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9OXipUDASOncyDee)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1214)):</sup>
Maybe `tag_map`?
<!-- Reviewable comment -K1f9OXjn1uBlsbpDBdk:889095891 -->

---

<sup>**[components/script/dom/document.rs, line 1428 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f8wsp6obm7VgOqOG5)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1428)):</sup>
Consider using `HashMap::entry` instead of `get`.  (Also in other places.)
<!-- Reviewable comment -K1f8wsp6obm7VgOqOG6:-2140915176 -->

---

<sup>**[components/script/dom/document.rs, line 1433 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f98tPbDgJbn3M54oa)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1433)):</sup>
Whitespace.
<!-- Reviewable comment -K1f98tPbDgJbn3M54ob:471607052 -->

---

<sup>**[components/script/dom/document.rs, line 1452 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1fB4Lu99NYywI6o_Iy)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1452)):</sup>
Whitespace.
<!-- Reviewable comment -K1fB4LvXAUY8z4XfHVx:471607052 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 30 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9cbJ6ijzDHs2KbMF)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L30)):</sup>
Consider using u64 for version?  It's easy to overflow a u32 on a modern processor.
<!-- Reviewable comment -K1f9cbKm5SuwuAU-QPu:775434527 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 171 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1fAAhEvj7UvXpW8q6W)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L171)):</sup>
Whitespace.
<!-- Reviewable comment -K1fAAhFb0CipJ2tEOv0:471607052 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 282 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1fATamLjCHYJRWjbaV)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L282)):</sup>
Whitespace.
<!-- Reviewable comment -K1fATamLjCHYJRWjbaW:471607052 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Thanks! I'll have a look at your review comments tomorrow. Also get test-tidy to pass, sigh.

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/document.rs, line 1214 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9OXipUDASOncyDee)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1214)):</sup>
The name was just for consistency with idmap. I'm happy to change all of them.
<!-- Reviewable comment -K1fwqPV42emJvDQy9NG:1522214744 -->

---

<sup>**[components/script/dom/document.rs, line 1428 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f8wsp6obm7VgOqOG5)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1428)):</sup>
Good point. I'm still learning my way round the rust stdlib.
<!-- Reviewable comment -K1fxCSF2L0s6tlm1W6R:43722593 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 30 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9cbJ6ijzDHs2KbMF)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L30)):</sup>
Tradeoff of space vs correctness? The u32 is only a problem if there's an overflow and a version comparison with exactly the right old version. The u64 uses an extra word per DOM node.
<!-- Reviewable comment -K1fxOqUTnjh8ZTJ9ge4:2018382276 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmlcollection.rs, line 30 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9cbJ6ijzDHs2KbMF)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L30)):</sup>
I'm not convinced saving a word is worth compromising correctness... but if you really want it to wrap, at least use `Wrapping<u32>` so it doesn't panic in overflow-checked builds.
<!-- Reviewable comment -K1fyXLci_rYtDTNbriA:804024808 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmlcollection.rs, line 30 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f9cbJ6ijzDHs2KbMF)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/htmlcollection.rs#L30)):</sup>
On reflection, I think the space saving probably isn't worth it, I'll change this to u64.
<!-- Reviewable comment -K1iOLTSpypYad43-v2Z:695048140 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/document.rs, line 1428 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f8wsp6obm7VgOqOG5)** ([raw file](https://github.com/servo/servo/blob/4dafec1719be8e4d2405f033b33c59b571d9626a/components/script/dom/document.rs#L1428)):</sup>
On looking at the code again, I don't think I can use HashMap::entry, because of rooting.
<!-- Reviewable comment -K1ihXBqWQI1ZL28zZHR:766259020 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Because of rooting?  Hmm... that sounds like we just need to update the whitelist in the lint plugin; `std::collections::hash_map::Entry` should always be fine to store on the stack.  (https://github.com/servo/servo/blob/master/components/plugins/lints/unrooted_must_root.rs#L53 .)

asajeffrey commented:
The problem isn't that Entry needs rooted, it's that the code would end up being something like:

  self.foo_map.entry(key).or_insert(|| mk_foo().r() ).root()

that is, in the case of a cache miss, we build a rooted result, then unroot it, then root it again.

This is because the cache is being kept unrooted: should the cache be the rooted versions?
<!-- Reviewable comment -K1jelwqBnkeDdGYxEJW:-1335965331 -->

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
That doesn't seem like a big deal; rooting isn't an expensive operation.

If you really want to avoid it, you can do `match foo_map.entry(key) { VacantEntry(e) => {}`...
<!-- Reviewable comment -K1jg_HASQ-pZTWbaGDc:-1289314576 -->

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

metajack commented:
I believe that @asajeffrey has shown that rooting is expensive. It's certianly showing up in his profiles from what I understand.

asajeffrey commented:
The problem with rooting is that it uses a Vec for roots, so the code includes the case where the vector has to grow, which means problems for code inlining. It's more of a code size and optimization problem than a time problem, since in the common case rooting is just a vector push into a vector which already has enough backing store.

I'll update to using HashMap::entry.
<!-- Reviewable comment -K1jl-TtEJYmoj4kDDJq:2003698571 -->

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

tschneidereit commented:
@bors-servo: try

Manishearth commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 89248e8 with merge 4554d8a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/114)

Manishearth commented:
(Artificial failure)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8262) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
I responded to @eefriedman and rebased to builld with master. Anything else?

jdm commented:
@eefriedman Are you doing a full review of this, or just drive-by? Please set the assignee field for the former, so it's clear who's responsible for the next step here.

eefriedman commented:
I'll do a full review... or at least attempt it.  (I'll reassign if necessary.)

asajeffrey commented:
Thanks!
On Nov 2, 2015 2:31 PM, "eefriedman" <notifications@github.com> wrote:

> I'll do a full review... or at least attempt it. (I'll reassign if
> necessary.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/servo/servo/pull/8227#issuecomment-153147810>.
>

eefriedman commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 32 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29ZUu1qDBRyonwSg7O)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L32)):</sup>
You could formalize this by introducing a specific type with these semantics which can be converted to Option<u32>.  I'd like to see this because the cache code is confusing enough without having to remember weird special-cases.
<!-- Reviewable comment -K29ZUu2hMQB0Q1-DB6o:1774575530 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 126 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29f3Ia_UeGEdbZgdr_)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L126)):</sup>
`cached_length - index` can overflow.
<!-- Reviewable comment -K29f3IbUnfV_WL_rNwu:-1698198027 -->

---

<sup>**[components/script/dom/node.rs, line 335 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29bexwuyhnrqpT8_20)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L335)):</sup>
`self.owner_doc()` and `child.owner_doc()` are the same node; it would be more clear if you just computed the document once.
<!-- Reviewable comment -K29bexwuyhnrqpT8_21:2059227669 -->

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
It isn't obvious why we need to track the version of the current node separately from the version of its descendants.
<!-- Reviewable comment -K29gPJccOMb-2X-0wQr:654435540 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
This iterator is supposed to return nodes in reverse tree order, right?  A comment to that effect would be nice.  And the implementation is completely broken; I'd like to see a unit-test for this.
<!-- Reviewable comment -K29cqI3cXBdNVo1_CTK:1220861232 -->

---

<sup>**[tests/wpt/mozilla/tests/mozilla/htmlcollection.html, line 63 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29aeIJW14X863HTkvk)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/tests/wpt/mozilla/tests/mozilla/htmlcollection.html#L63)):</sup>
Trailing whitespace and tabs in this file.
<!-- Reviewable comment -K29aeIKSo2qyiV_ZmxN:-1502631754 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Thanks, I'll have a look at this tomorrow!
On Nov 2, 2015 18:20, "eefriedman" <notifications@github.com> wrote:

> Review status: 0 of 6 files reviewed at latest revision, 6 unresolved
> discussions.
> ------------------------------
>
> *components/script/dom/htmlcollection.rs, line 32 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29ZUu1qDBRyonwSg7O>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L32>
> ):
> You could formalize this by introducing a specific type with these
> semantics which can be converted to Option. I'd like to see this because
> the cache code is confusing enough without having to remember weird
> special-cases.
> ------------------------------
>
> *components/script/dom/htmlcollection.rs, line 126 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29f3Ia_UeGEdbZgdr_>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L126>
> ):
> cached_length - index can overflow.
> ------------------------------
>
> *components/script/dom/node.rs, line 335 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29bexwuyhnrqpT8_20>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L335>
> ):
> self.owner_doc() and child.owner_doc() are the same node; it would be
> more clear if you just computed the document once.
> ------------------------------
>
> *components/script/dom/node.rs, line 545 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545>
> ):
> It isn't obvious why we need to track the version of the current node
> separately from the version of its descendants.
> ------------------------------
>
> *components/script/dom/node.rs, line 1172 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172>
> ):
> This iterator is supposed to return nodes in reverse tree order, right? A
> comment to that effect would be nice. And the implementation is completely
> broken; I'd like to see a unit-test for this.
> ------------------------------
>
> *tests/wpt/mozilla/tests/mozilla/htmlcollection.html, line 63 [r5]
> <https://reviewable.io:443/reviews/servo/servo/8227#-K29aeIJW14X863HTkvk>*
> (raw file
> <https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/tests/wpt/mozilla/tests/mozilla/htmlcollection.html#L63>
> ):
> Trailing whitespace and tabs in this file.
> ------------------------------
>
> Comments from the review on Reviewable.io
> <https://reviewable.io:443/reviews/servo/servo/8227>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/servo/servo/pull/8227#issuecomment-153201972>.
>

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 32 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29ZUu1qDBRyonwSg7O)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L32)):</sup>
Okay, that's a good idea.
<!-- Reviewable comment -K2E3fi1fFrHC7PATRX8:-1564262413 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 126 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29f3Ia_UeGEdbZgdr_)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L126)):</sup>
Do you mean underflow? If so, yes, I need to guard for the case where index > length.
<!-- Reviewable comment -K2E4KnenR0q1JRRho81:1449306407 -->

---

<sup>**[components/script/dom/node.rs, line 335 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29bexwuyhnrqpT8_20)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L335)):</sup>
This is future-proofing in case this is called in the middle of adopting a node, where the child node may be owned by a different document.
<!-- Reviewable comment -K2E4sntg7-kdpvdf4OG:-1482029581 -->

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
This is future-proofing in case there's other uses of versioning.
<!-- Reviewable comment -K2E5J9SrC8GVLyFt86e:-1342197687 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
It's the same as the following node iterator, but in reverse document order. It's the same as FollowingNodeIterator, but with left and right swapped (so GetFirstChild becomes GetLastChild and GetNextSibling becomes GetPreviousSibling). I don't see why this is broken.

The DOM library doesn't support unit tests. It would be good to add some more tests to tests/wpt/mozilla/tests/mozilla/htmlcollection.html though.
<!-- Reviewable comment -K2E5ytnSe7wn4XkTMQ1:212625017 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 126 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29f3Ia_UeGEdbZgdr_)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L126)):</sup>
Yes, that's what I mean.  (The term underflow generally isn't used for integer arithmetic.)
<!-- Reviewable comment -K2E7njC4fs9MFtHMYa5:765359523 -->

---

<sup>**[components/script/dom/node.rs, line 335 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29bexwuyhnrqpT8_20)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L335)):</sup>
Okay.
<!-- Reviewable comment -K2E8XP3yf2sn822SWo5:-489118315 -->

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
Could you give an example of how you might use this?
<!-- Reviewable comment -K2E8Zl4WcHYAtL-fbkv:606623780 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
You can't unittest DOM nodes... but that doesn't mean you can't unit-test this algorithm.  Just make it generic over the node type, and you can mock up an alternative tree.

As for why this is wrong... take the tree `<a><b><c></c></b></a>`.  Tree order is `a`, `b`, `c`.  Reverse tree order is `c`, `b`, `a`.  Your algorithm returns `a`, `b`, `c`.
<!-- Reviewable comment -K2E907IqvpTFh6SMMvo:1476311067 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
I was thinking of examples where you want to cache a value that's computed just from the current node. An example might be the parsed CSS for a style attribute, which you'd invalidate when the node changes, or a cached child collection.
<!-- Reviewable comment -K2EIoANXKQpLsX2Jgcl:-49985175 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
If we're going to make this algorithm generic in the tree so it can be unit tested, shouldn't we do the same to the other iterators, and in general to the whole DOM navigation and query API?

Oh, I see what you mean about the error case. Back to the drawing board on that one. I can at least add an integration test for it.
<!-- Reviewable comment -K2EK2HKD5EpjwtqT8yI:1994171087 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
That doesn't really sound worthwhile: if we're caching something which doesn't depend on the children, we can usually clear the cache when the node is dirtied.  The reason we're using versions for collections is that iterating through all the cached collections and updating them would be expensive.
<!-- Reviewable comment -K2ENxwshnSncEn4oQB3:-1922793296 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
For unit-testing, making all the iterators generic would be nice, yes. Probably out of scope for this PR, though.
<!-- Reviewable comment -K2EPtiDlWaHUki95nM1:206038046 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8221) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
Yes, but that means using a push mechanism rather than pull. That said, it's just future-proofing, so I can remove it for this PR.
<!-- Reviewable comment -K2E_EuQ-2Fc-uOnCQ3_:2074661928 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 126 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29f3Ia_UeGEdbZgdr_)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/htmlcollection.rs#L126)):</sup>
Fixed.
<!-- Reviewable comment -K2HzBXBzmReWrWHcYOq:1758359711 -->

---

<sup>**[components/script/dom/node.rs, line 545 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29gPJccOMb-2X-0wQq)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L545)):</sup>
I removed it.
<!-- Reviewable comment -K2HybocuB1kUuoOteRS:-2096001433 -->

---

<sup>**[components/script/dom/node.rs, line 1172 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8227#-K29cqI3cXBdNVo1_CTJ)** ([raw file](https://github.com/servo/servo/blob/b28638ac4cbb790dbb7aa717405e3820ef13bdef/components/script/dom/node.rs#L1172)):</sup>
I removed this custom iterator, and just used the built-in node.preceding_nodes(root) iterator instead.

I added a regression test for this bug.
<!-- Reviewable comment -K2HzQrBaq3JjmlqtdYM:284430560 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
Reviewed 1 of 5 files at r2, 2 of 5 files at r6, 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 30 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8227#-K2IGqrTUMOxcm1j5_MG)** ([raw file](https://github.com/servo/servo/blob/3c7ef3f65784b9d5e4e1371012bd5d3d0c34a485/components/script/dom/htmlcollection.rs#L30)):</sup>
Probably better to use a more descriptive name, like `OptionalU32`.
<!-- Reviewable comment -K2IGqrTUMOxcm1j5_MH:-1684567502 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 44 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8227#-K2IGy-lyaE8ulSXLV_V)** ([raw file](https://github.com/servo/servo/blob/3c7ef3f65784b9d5e4e1371012bd5d3d0c34a485/components/script/dom/htmlcollection.rs#L44)):</sup>
This should probably assert that `bits != u32::max_value()`.
<!-- Reviewable comment -K2IGy-lyaE8ulSXLV_W:-61874961 -->

---

<sup>**[components/script/dom/htmlcollection.rs, line 61 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8227#-K2IH9-z52Z8TTwlajF_)** ([raw file](https://github.com/servo/servo/blob/3c7ef3f65784b9d5e4e1371012bd5d3d0c34a485/components/script/dom/htmlcollection.rs#L61)):</sup>
Sorry I didn't think of this earlier... but could this lead to a "leak"?  For example, a subtree is removed from the document, but doesn't get garbage-collected until the document itself is collected.

(If the answer here is "we need weak references", it's fine to add a FIXME and leave it for now.)
<!-- Reviewable comment -K2IH90-XYnYQeTpapWx:-2121102336 -->

---

<sup>**[tests/wpt/mozilla/tests/mozilla/htmlcollection.html, line 63 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8227#-K2IIb2s57gQawW97wr-)** ([raw file](https://github.com/servo/servo/blob/3c7ef3f65784b9d5e4e1371012bd5d3d0c34a485/tests/wpt/mozilla/tests/mozilla/htmlcollection.html#L63)):</sup>
Trailing whitespace.
<!-- Reviewable comment -K2IIb2s57gQawW97wr0:-243822606 -->

---

<sup>**[tests/wpt/mozilla/tests/mozilla/htmlcollection.html, line 101 \[r9\]](https://reviewable.io:443/reviews/servo/servo/8227#-K2IIcjV2RmwXjl6hzuS)** ([raw file](https://github.com/servo/servo/blob/3c7ef3f65784b9d5e4e1371012bd5d3d0c34a485/tests/wpt/mozilla/tests/mozilla/htmlcollection.html#L101)):</sup>
Tabs.
<!-- Reviewable comment -K2IIcjV2RmwXjl6hzuT:-484797845 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

asajeffrey commented:
OK, I'll make these changes tomorrow. Hopefully this will be the last round!

asajeffrey commented:
Changes added. Thanks for all your hard work @eefriedman !

eefriedman commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 8299712 has been approved by `eefriedman`

<!-- @bors-servo r=eefriedman 8299712e63d037ba2b767b6bf42d2d06adb32bfa -->

bors-servo commented:
:hourglass: Testing commit 8299712 with merge 7ec64b8...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/275)

eefriedman commented:
This is causing unittest failures because you changed the size of misc DOM nodes.  Please update the unit-tests...

CC @Ms2ger to double-check that this is okay.

asajeffrey commented:
OK, I guess we need input about adding one word of storage per DOM node. I'll ask #servo.

metajack commented:
Just to capture the IRC flavor, it was agreed this word was worth it. The check is there to prevent accidentally increasing the size since that used to happen. If there's a reason to increase it, it is generally fine.

cc @pcwalton

eefriedman commented:
Oh, one more thing: please squash the commits.

asajeffrey commented:
@eefriedman are you okay with me just squashing the fixup commit, or is there more squashing you'd like to see?

eefriedman commented:
Convention in Servo is to squash everything.  (Occasionally, separate commits make the history more readable, so we'll keep them around... but that's the exception, not the rule.)

jdm commented:
I don't think the convention is actually to squash everything into single commits. We encourage that when the commits are largely not made independent of each other, but we also encourage smaller independent commits that can compile in isolation.

jdm commented:
Basically, if not squashing commits can make it easier to understand the changes later, there's no reason to require it.

aravind-pg commented on components/script/dom/node.rs:
```diff
@@ -105,6 +106,9 @@ pub struct Node {
     flags: Cell<NodeFlags>,
 
+    /// The maximum version of any inclusive descendent of this node.
+    inclusive_descendents_version: Cell<u64>,
```
Is this deliberately "descendents" instead of "descendants", here and elsewhere? I believe the latter is much more standard.

asajeffrey replied:
Oops. Good catch.

Commit:
Made QualName traceable by JS.

Commit:
Added versioning to DOM nodes.

There is now an inclusive_descendants_version field of each node, which
increases each time the node, or any of its descendants, is dirtied.
This can be used for cache invalidation, by caching a version number
and comparting the current version number against the cached version number.

Commit:
Caching HTMCollections.

We cache the state of any live HTMLCollection, keeping track of

a) the optional cached length of the collection, and
b) an optional cursor into the collection (a node in the collection plus its index).

The cache is invalidated based on the version number of the node.

We use these caches for speeding up random access to the collection.
When returning coll[i], we search from the cursor, if it exists,
and otherwise search from the front of the collection.
In particular, both a forward for-loop and a backward for-loop
through the collection will now have each access take O(1)
time rather than O(n) time.

This gets 1000x speed-up on the relevant Dromaeo DOM query tests.

asajeffrey commented:
Squashed down to 3 commits. Fixed spelling mistake. Hopefully that's it!

aravind-pg commented on components/script/dom/htmlcollection.rs:
```diff
@@ -6,76 +6,169 @@ use dom::bindings::codegen::Bindings::HTMLCollectionBinding;
     #[ignore_heap_size_of = "Contains a trait object; can't measure due to #6870"]
-    collection: Collection,
+    filter: Box<CollectionFilter + 'static>,
+    // We cache the version of the root node and all its decendents,
```
decendents --> descendants :)

asajeffrey commented:
Review status: 2 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[components/script/dom/htmlcollection.rs, line 58 \[r17\]](https://reviewable.io:443/reviews/servo/servo/8227#-K1f2sTi12xunH3JzZvJ-r17-58)** ([raw file](https://github.com/servo/servo/blob/3e4cc155af688950379bc1521857642aec72b7ed/components/script/dom/htmlcollection.rs#L58)):</sup>
Fixed.
<!-- Reviewable comment -K2SmlD885fpKzfTqJly:1758359711 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8227)
<!-- Sent from Reviewable.io -->

eefriedman commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 3e4cc15 has been approved by `eefriedman`

<!-- @bors-servo r=eefriedman 3e4cc155af688950379bc1521857642aec72b7ed -->

bors-servo commented:
:hourglass: Testing commit 3e4cc15 with merge d201bbe...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/293)

eefriedman commented:
```
Running 6693 tests in web-platform-tests

  ▶ PASS [expected FAIL] /css21_dev/html4/block-in-inline-remove-000.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/block-in-inline-remove-002.htm
```

(On both Linux and Mac.)

eefriedman commented:
I'll try to merge #8315, and then you can rebase.

asajeffrey commented:
Thanks! I'll rebase... Grumble mutter dependent PRs.

eefriedman commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 237ddc3 has been approved by `eefriedman`

<!-- @bors-servo r=eefriedman 237ddc3c0cdf1d15941a25f6df18ea76a309d860 -->

bors-servo commented:
:hourglass: Testing commit 237ddc3 with merge 9dbd5dd...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/348)

eefriedman commented:
@bors-servo retry #8416

bors-servo commented:
:hourglass: Testing commit 237ddc3 with merge f39faaf...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/342), [gonk](http://build.servo.org/builders/gonk/builds/350), [linux-dev](http://build.servo.org/builders/linux-dev/builds/345), [linux-rel](http://build.servo.org/builders/linux-rel/builds/376), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/357), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/355), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/372)
